Upgrading explanation page (#950)
canonical/k8s-snap: docs/src/snap/explanation/upgrade.md

# Upgrading with {{product}}

Upgrading your cluster to the latest Kubernetes version is a critical part of
maintaining a healthy cluster. It provides enhanced security in the form
of the latest security patches as well as increased stability and performance
by incorporating the latest optimizations and features. The [release cadence]
of {{product}} follows the [release cycle] set by upstream Kubernetes.

## Patch vs minor upgrades

Patch upgrades address bug fixes and are typically safe, introducing no
breaking changes.
Patch upgrades do not change the channel selected (for example a patch upgrade
from stable/1.32.0 to stable/1.32.1). Minor versions upgrades add new
features or deprecate existing features without breaking changes. These upgrades
change the version of the channel selected
(for example a minor upgrade from stable/1.32.x to stable/1.33.x).
For more information on channels see the [explanation page].

## Sequential upgrades

Upgrades must always be done sequentially. It is not recommended to upgrade by

> more than one minor version at a time (for example upgrading from 1.31 to 1.33

Reviewer: This only applies for control-planes, workers can have a version skew of up to 3

is not supported). Changing between minor releases (for example upgrading
from 1.31 to 1.32) is also only recommended when you are on the
latest patch of the current version (for example 1.31.x where x is the latest
available patch in 1.31 channel). This is in line with upstream Kubernetes
[version skew policy].

## Upgrade approach

It is important to upgrade your cluster with the same method you used to install
the cluster. If you have deployed {{product}} using Juju, you must refresh
using Juju controller commands. Similarly with snap and CAPI deployments,
the original install method must be used.

## Snap upgrades

Updates for the {{product}} snap are checked several times a day with the

> default [snap refresh fucntionality]. Patch upgrades will be applied

Reviewer 2: ```suggestion
default [snap refresh functionality]. Patch upgrades will be applied
```

automatically unless they are manually stopped. Changes applied during the
latest refresh to the snap can be viewed using the command `snap change`.
Minor version upgrades need manual intervention and must be carried out on all
snaps individually. The length of time taken to upgrade the snap will depend
largely on your local setup, the services running on the cluster, resources
available to the cluster,the intensity of the workloads etc.

If you would like to manage the upgrading of your cluster using the snap please
see the how-to guide on [managing cluster upgrades].

## Charm upgrades

Upgrading the charm must be done manually for both patch and minor upgrades on
a per model basis. To ensure that the revision of the charm always matches the

> revision of the snap it orchestrates, the `refresh` function of the snap has

Reviewer: I would rephrase this. 
For me, this sounds like one cannot execute a snap refresh on a Juju unit (which is not true)
```
ubuntu@juju-691454-1:~$ sudo snap refresh k8s --channel=latest/edge
2025-01-13T09:21:16Z INFO Waiting for "snap.k8s.kube-apiserver.service" to stop.
k8s (edge) v1.32.0 from Canonical✓ refreshed
```

Maybe something like this fits better?

The charm locks the k8s snap to a specific revision, preventing automatic upgrades. This ensures the snap version remains in sync with the charm version, and all upgrades are managed exclusively through Juju.

Reviewer 2: Actually, the snap refresh mechanism is on hold until the charm logic is instructed to upgrade the snap. Maybe we can include some of this information: https://snapcraft.io/docs/managing-updates#p-32248-if-snaps-are-specified

Reviewer: I'm not sure if I understand your point @mateoflorido ?
Could you please elaborate how this differs from my suggestion? Thanks!

Author: I have expanded this a bit more. Does it make more sense now?

been frozen and all upgrades must be done through Juju. The `juju refresh`
command instructs the Juju controller
to use the new charm revision within the current charm channel or to upgrade to
the next channel based on the parameters provided. The charm code
is simultaneously replaced on each unit. Then, the k8s snap is updated
unit-by-unit. This is in order to maintain a highly-available kube-api-server
endpoint, starting with the Juju leader unit for the application. To ensure a
smooth upgrade process, the pre-upgrade-check should be run before
all upgrades. This ensures that you cluster is in a stable and ready state for

> an upgrade.

Reviewer 2: Perhaps we can rephrase this. The `pre-upgrade-check` is mandatory before upgrading a cluster. In addition to verifying the cluster's health, it also creates the upgrade stack which will be used after the charm is upgraded to orchestrate the upgrade process across the target application (`k8s` or `k8s-worker`) units

If you would like to upgrade your cluster using Juju please see the how-to guide
on [upgrading your cluster by a minor version].

> <!-- TODO CAPI Upgrades-->

Reviewer: Do we have a follow-up card for this? Why don't we add this as part of this PR?

Author: We have not fully developed/agreed upon the upgrade story for CAPI so we cannot document it yet. I will add a card to complete this page when possible. Thanks for reminding me.

> ## High availability with cluster upgrades

Reviewer: Should we link to the dedicated HA page in this paragraph for further information?

Reviewer 3: +1 I would try to avoid duplication and minimize this section.

not include any customized configuration of Kubernetes, or no-build-in
generated objects (e.g. storage classes) or deployments which rely on
deprecated APIs.

## Updating a cluster best practices

When upgrading with {{product}} best practices should be followed:

- Check the release notes associated with the release and pay
particular attention to any deprecations or incompatible changes being
introduced that may affect your cluster.
- Create a backup of all the information in the cluster.
- Upgrade the control plane nodes before upgrading the worker nodes.
- Cluster administrators should identify critical workloads on the node to be
updated and decide if the node needs to be cordoned and drained before cluster
upgrade.
- Deploy a rolling upgrade strategy meaning you upgrade one node successfully
to completion before upgrading the next node.

<!-- LINKS -->
[release cadence]: https://ubuntu.com/about/release-cycle#canonical-kubernetes-release-cycle
[release cycle]: https://kubernetes.io/releases/release/
[managing cluster upgrades]: ../howto/upgrades
[upgrading your cluster by a minor version]: ../../charm/howto/upgrade-minor/

> [snap refresh fucntionality]:https://snapcraft.io/docs/refresh-awareness

Reviewer 2: ```suggestion
[snap refresh functionality]:https://snapcraft.io/docs/refresh-awareness
```

> # Upgrading with {{product}}

Reviewer 3: ```suggestion
# Upgrading {{product}}
```

Upgrading your cluster to the latest Kubernetes version is a critical part of
maintaining a healthy cluster. It provides enhanced security in the form
of the latest security patches as well as increased stability and performance
by incorporating the latest optimizations and features. The [release cadence]
of {{product}} follows the [release cycle] set by upstream Kubernetes.

## Patch vs minor upgrades

Patch upgrades address bug fixes and are typically safe, introducing no
breaking changes.
Patch upgrades do not change the channel selected (for example a patch upgrade

> from stable/1.32.0 to stable/1.32.1). Minor versions upgrades add new

Reviewer 3: ```suggestion
from stable/1.32.0 to stable/1.32.1). Minor version upgrades add new
```
Maybe?

features or deprecate existing features without breaking changes. These upgrades
change the version of the channel selected
(for example a minor upgrade from stable/1.32.x to stable/1.33.x).

> For more information on channels see the [explanation page].

Reviewer 3: Should we mention any upgrades for CVE & bug fixes on our end and recommend not pinning the snap?
(This could even be later in this doc)

Author: Added to the snap paragraph

## Sequential upgrades

Upgrades of control plane nodes
must always be done sequentially. It is not recommended to upgrade by
more than one minor version at a time (for example upgrading from 1.31 to 1.33

> is not supported). Changing between minor releases (for example upgrading

Reviewer 3: ```suggestion
is **not** supported). Changing between minor releases (for example upgrading
```

> is not supported). Changing between minor releases (for example upgrading
> from 1.31 to 1.32) is also only recommended when you are on the
> latest patch of the current version (for example 1.31.x where x is the latest
> available patch in 1.31 channel). This is in line with upstream Kubernetes

Reviewer 3: ```suggestion
is not supported). 
When upgrading between minor releases, for instance from 1.31 to 1.32, first upgrade to the
latest patch of the current version. In this example you would check the 1.31 channel for the latest patch: 1.31.x. This is in line with upstream Kubernetes
```
Consider splitting up this sentence

[version skew policy]. For worker nodes, upgrading
with a skew of up to three versions is acceptable.

## Upgrade approach

> It is important to upgrade your cluster with the same method you used to install
> the cluster. If you have deployed {{product}} using Juju, you should refresh

Reviewer 3: ```suggestion
It is important to upgrade your cluster with the same method that you used to install
the cluster. If you have deployed {{product}} using Juju, you should upgrade your cluster
```

> using Juju controller commands. Similarly with snap and CAPI deployments,
> the original install method should be used.

Reviewer 3: ```suggestion
using Juju controller commands. Similarly, with snap and CAPI deployments,
the original install method should be used.
```
I'm no comma wizard but is this one required?

Author: I think both are grammatically correct - it is just a matter of style. I'll add it anyways.

## Snap upgrades

Updates for the {{product}} snap are checked several times a day with the
default [snap refresh functionality]. Patch upgrades will be applied
automatically unless they are manually stopped. Changes applied during the
latest refresh to the snap can be viewed using the command `snap change`.
Minor version upgrades need manual intervention and must be carried out on all
snaps individually. The length of time taken to upgrade the snap will depend

> largely on your local setup, the services running on the cluster, resources
> available to the cluster,the intensity of the workloads etc.

Reviewer 3: ```suggestion
largely on your local setup, the services running on the cluster, the resources
available to the cluster and the intensity of the workloads etc.
```

If you would like to manage the upgrading of your cluster using the snap, please
see the how-to guide on [managing cluster upgrades].

## Charm upgrades

Upgrading the charm must be done manually for both patch and minor upgrades on
a per model basis. To ensure that the revision of the charm matches the
revision of the snap it orchestrates, the automatic `refresh` function of the
snap has been placed on hold in this deployment method. The snap can still be
manually updated by running a [targeted snap refresh], but all upgrades should
be done through Juju. The `juju refresh`command instructs the Juju controller
to use the new charm revision within the current charm channel or to upgrade to
the next channel based on the parameters provided. The charm code
is simultaneously replaced on each unit. Then, the k8s snap is updated

> unit-by-unit. This is in order to maintain a highly-available kube-api-server

Reviewer 3: We can link our new HA page here!

endpoint, starting with the Juju leader unit for the application. To ensure a
smooth upgrade process, the mandatory pre-upgrade-check is run before
all upgrades. In addition to verifying the cluster's health,
the pre-upgrade-check also creates the upgrade stack which
will be used after the charm is upgraded. The upgrade stack is
used to orchestrate the upgrade process across the k8s or k8s-worker units.

If you would like to upgrade your cluster using Juju please see the how-to guide
on [upgrading your cluster by a minor version].

<!-- TODO CAPI Upgrades-->

## High availability with cluster upgrades

High availability is automatically enabled in {{ product }} for clusters with
three or more nodes independent of the deployment method. Clusters without high

> availability must be extra vigilant on backing up cluster data before starting

Reviewer 3: Do we have any cluster backup docs or can we point to external tools?

the upgrade process and also must be aware of potential service disruptions
during cluster upgades. It is also important to understand that Kubernetes will
only upgrade and if necessary migrate, components of Kubernetes installed and
configured as part of Kubernetes. This may
not include any customized configuration of Kubernetes, or no-build-in
generated objects (e.g. storage classes) or deployments which rely on
deprecated APIs. To find out more about high availability in a {{product}}
cluster see the [high availability explanation page].

## Updating a cluster best practices

When upgrading with {{product}} best practices should be followed:

> - Check the release notes associated with the release and pay

Reviewer 3: A link to the release notes please.
